lxd/device/device.go

```diff
@@ -0,0 +1,128 @@
+	// either due to unplugging it from a running instance or instance is being shutdown.
+	Stop() error
+
+	// Remove performs any host-side cleanup when an instance is removed from an instance.
+	Remove() error
+}
+
+// device represents a sealed interface that implements Device, but also contains some internal
+// setup functions for a Device that should only be called by device.New() to avoid exposing devices
+// that are not in a known configured state. This is separate from the Device interface so that
+// Devices created outside of the device package can be used by LXD, but ensures that any devices
+// created by the device package will only be accessible after being configured properly by New().
+type device interface {
+	Device
+
+	// init stores the InstanceIdentifier, daemon State and Config into device and performs any setup.
+	init(InstanceIdentifier, *state.State, config.Device, func() map[string]string, func(map[string]string) error)
+
+	// validate checks Config stored by init() is valid for the instance type.
+	validate() error
+}
+
+// deviceCommon represents the common struct for all devices.
+type deviceCommon struct {
+	instance    InstanceIdentifier
+	config      map[string]string
+	state       *state.State
+	volatileGet func() map[string]string
+	volatileSet func(map[string]string) error
+}
+
+// init stores the InstanceIdentifier, daemon state and Config into device. It can also be provided
+// with volatile get and set functions for the device to allow persistent data to be accessed.
+// This is implemented as part of deviceCommon so that the majority of devices don't need to
+// implement it and can just embed deviceCommon.
+func (d *deviceCommon) init(instance InstanceIdentifier, state *state.State, conf config.Device, volatileGet func() map[string]string, volatileSet func(map[string]string) error) {
+	d.instance = instance
+	d.config = conf
+	d.state = state
+	d.volatileGet = volatileGet
+	d.volatileSet = volatileSet
+}
+
+// Add returns nil error as majority of devices don't need to do any host-side setup.
+func (d *deviceCommon) Add() error {
+	return nil
+}
+
+// CanHotPlug returns true as majority of devices can be started/stopped when instance is running.
+// Also returns an empty list of update fields as most devices do not support live updates.
+func (d *deviceCommon) CanHotPlug() (bool, []string) {
+	return true, []string{}
+}
+
+// Update returns an error as most devices do not support live updates without being restarted.
+func (d *deviceCommon) Update(oldConfig config.Device, isRunning bool) error {
+	return fmt.Errorf("Device does not support updates whilst started")
+}
+
+// Remove returns nil error as majority of devices don't need to do any host-side cleanup on delete.
+func (d *deviceCommon) Remove() error {
+	return nil
+}
+
+// New instantiates a new device struct, validates the supplied config and sets it into the device.
```

Comment:
To improve the generated documentation you should either comment on what volatileGet and volatileSet do, or, preferably have both of them defined as exported types at the top of the file with their individual descriptions explaining what's expected from those.

Reply:
@stgraber sounds good, will add.